### PR TITLE
Helptickets: Refactor to support text-only tickets

### PR DESCRIPTION
**Pull request by mia-pi-git**
This should cover support for almost everything discussed here https://www.smogon.com/forums/threads/3673339/page-2#post-8677283.
I opted for a separate resolving command for text tickets to avoid cluttering up the other command, though /ht close will parse /ht resolve if the target ticket is a text ticket.

### CODE REVIEW

**Comment by HoeenCoder** [comment]
I just want to mention I have seen the review request and will review but am going to be busy for a few days, sorry about that.

File: server/chat-plugins/helptickets.ts [hunk 1]
```diff
@@ -558,6 +601,17 @@ function checkIp(ip: string) {
 	return false;
 }
 
+export function getBattleLinks(text: string) {
+	const rooms: string[] = [];
+	const battles = text.match(BATTLES_REGEX);
+	// yes this is necessary, exec has behavior here that makes it inferior
```
Comment (Zarel) [inline server/chat-plugins/helptickets.ts:607]:
```suggestion
	// typescript-eslint is having trouble detecting REPLAY_REGEX as a global regex
```

File: server/chat-plugins/helptickets.ts [hunk 2]
```diff
@@ -1,10 +1,16 @@
...
 import type {Punishment} from '../punishments';
+import type {PartialModlogEntry, ModlogID} from '../modlog';
 
 const TICKET_FILE = 'config/tickets.json';
 const TICKET_CACHE_TIME = 24 * 60 * 60 * 1000; // 24 hours
 const TICKET_BAN_DURATION = 48 * 60 * 60 * 1000; // 48 hours
+const BATTLES_REGEX = /battle-(?:[a-z0-9]+)-(?:[0-9]+)(?:-[a-z0-9]+pw)?/g;
+const REPLAY_REGEX = new RegExp(
+	`${Config.routes.replays}` +
+	`/(?:[a-z0-9]-)?(?:[a-z0-9]+)-(?:[0-9]+)(?:-[a-z0-9]+pw)?`, "g"
```
Comment (Zarel) [inline server/chat-plugins/helptickets.ts:12]:
```suggestion
	`${Config.routes.replays}/(?:[a-z0-9]-)?(?:[a-z0-9]+)-(?:[0-9]+)(?:-[a-z0-9]+pw)?`, "g"
```

Putting this on two lines is kind of mysterious, though

File: server/chat-plugins/helptickets.ts [hunk 3]
```diff
@@ -636,6 +697,134 @@ const cheatingScenarios = [
...
+
+			let battlelogNoticeAdded = false;
+			for (const [i, url] of rooms.entries()) {
+				if (rooms.indexOf(url) !== i) {
+					rooms.splice(i, 1);
+					continue;
+				}
+				const room = Rooms.get(url);
+				if (room) {
+					staff.joinRoom(room, connection);
```
Comment (HoeenCoder) [inline server/chat-plugins/helptickets.ts:752]: I'm not a fan of automatically joining battles when I take a ticket, especially for things like champ battles.

**Review by HoeenCoder — COMMENTED** [review]
Sorry for the delay, heres the review.

File: server/chat-plugins/helptickets.ts (hunk 3 above)
Comment (mia-pi-git) [reply server/chat-plugins/helptickets.ts:752]: Was explicitly requested. I can talk to Teclis about it.

File: server/chat-plugins/helptickets.ts [hunk 4]
```diff
@@ -408,6 +440,19 @@ export class HelpTicket extends Rooms.RoomGame {
...
+		if (seen) return;
+		user.send(`|pm|&Staff|${user.getIdentity()}|Hello! Your report was resolved by ${by}, ${Chat.toDurationString(Date.now() - time)} ago.`);
+		if (result?.trim()) {
+			user.send(`|pm|&Staff|${user.getIdentity()}|The result was "${result}"`);
+		}
+		tickets[userid].resolved!.seen = true;
+		writeTickets();
+	}
+	static getTypeId(name: string) {
+		return Object.entries(ticketTitles).find(entry => entry[1] === name)?.[0] || toID(name);
```
[the remark this answers is not in the document]
Comment (mia-pi-git) [reply server/chat-plugins/helptickets.ts:454]: That's how it works with optional chaining.

File: server/chat-plugins/helptickets.ts [hunk 5]
```diff
@@ -1,10 +1,15 @@
...
 import {getCommonBattles} from '../chat-commands/info';
 import type {Punishment} from '../punishments';
+import type {PartialModlogEntry, ModlogID} from '../modlog';
 
 const TICKET_FILE = 'config/tickets.json';
 const TICKET_CACHE_TIME = 24 * 60 * 60 * 1000; // 24 hours
 const TICKET_BAN_DURATION = 48 * 60 * 60 * 1000; // 48 hours
+const BATTLES_REGEX = /battle-(?:[a-z0-9]+)-(?:[0-9]+)(?:-[a-z0-9]+pw)?/g;
+const REPLAY_REGEX = new RegExp(
+	`${Config.routes.replays}/(?:[a-z0-9]-)?(?:[a-z0-9]+)-(?:[0-9]+)(?:-[a-z0-9]+pw)?`, "g"
```
[the remark this answers is not in the document]
Comment (mia-pi-git) [reply server/chat-plugins/helptickets.ts:11]: Probably. I'll check on that.

File: server/chat-plugins/helptickets.ts [hunk 6]
```diff
@@ -40,6 +56,9 @@ try {
 					const ticketGame = ticketRoom.game as HelpTicket;
 					ticketGame.writeStats(false);
 					ticketRoom.expire();
+				} else if (ticket.text) {
```
[the remark this answers is not in the document]
Comment (mia-pi-git) [reply server/chat-plugins/helptickets.ts:59]: Yeah, probably. I'll add that.

File: server/chat-plugins/helptickets.ts [hunk 7]
```diff
@@ -636,6 +697,134 @@ const cheatingScenarios = [
...
+			const rooms = getBattleLinks(text);
+
+			if (context) {
+				rooms.push(...getBattleLinks(context));
+			}
+
+			let battlelogNoticeAdded = false;
+			for (const [i, url] of rooms.entries()) {
+				if (rooms.indexOf(url) !== i) {
+					rooms.splice(i, 1);
```
[the remark this answers is not in the document]
Comment (mia-pi-git) [reply server/chat-plugins/helptickets.ts:747]: Alright. I'll just run this on a copy of the array so it doesn't affect the current loop.

File: server/chat-plugins/helptickets.ts (hunk 3 above)
Comment (HoeenCoder) [reply server/chat-plugins/helptickets.ts:752]: Yeah, me and Annika are concerned about it, I would assume other staff would be as well mostly due to champ battles.

Comment (mia-pi-git) [reply server/chat-plugins/helptickets.ts:752]: See #private-dev. We also really don't have that many champ reports anymore, I personally don't see it as much of an issue.

Comment (HoeenCoder) [reply server/chat-plugins/helptickets.ts:752]: Felucia also mentions its possibly exploitable by reporting battles with a ton of turns to freeze the browser.

Comment (mia-pi-git) [reply server/chat-plugins/helptickets.ts:752]: Fair enough.

Comment (mia-pi-git) [reply server/chat-plugins/helptickets.ts]: Done.

**Pull request merged**